base/experiments/ad9910_ram.py

```diff
@@ -0,0 +1,206 @@
+        """Generate a RAM profile to a DDS channel.
+            Include optional parameters based on the "Data Source Priority" of
+            AD9910. See the switch case below to find out which optional
+            parameter is needed.
+
+        Args:
+            dds: The DDS that will playback the RAM profile.
+            data: Data (amplitude, phase, frequencies) to be put into the RAM.
+            ramp_interval: The time interval between each step of the RAM.
+                Note: If possible, keep the interval at a multiple of
+                    4*T_sysclk. T_sysclk is generally 1ns.
+            ram_type: See the RAMType enum.
+            ramp_mode: The playback mode of the RAM. See urukul.py in ARTIQ.
+            base_frequency: (Optional) Unmodulated DDS frequency.
+            base_phase: (Optional) Unmodulated DDS phase.
+            base_amplitude: (Optional) Unmodulated DDS amplitude.
+
+        Raises:
+            NotImplementedError: Unsupported RAM types found.
+        """
+        # Make sure the RAM can hold the entire sequence
+        assert (len(data) <= RAMProfile.RAM_SIZE)
+
+        ram = np.zeros((len(data),), dtype=int)
+        # Avoid contaminating the passed in data list
+        data = data.copy()
+        # SPI transaction of AD9910 mandates the transmission of higher
+        # order words before lower order words
+        # The data is reversed such that the first word shows up first
+        data.reverse()
+
+        # Initialize default values for the FTW, POW, and ASF registers
+        # This is with accordance to set_mu() in AD9910
+        self.ftw, self.pow, self.asf = 0, 0, 0x3fff
+
+        # Encode the FTW, POW, ASF, and raw data en masse
+        match ram_type:
+            case RAMType.FREQ:
+                self.dest = RAM_DEST_FTW
+                dds.frequency_to_ram(data, ram)
+                self.asf = dds.amplitude_to_asf(base_amplitude)
+                self.pow = dds.turns_to_pow(base_phase)
+            case RAMType.PHASE:
+                self.dest = RAM_DEST_POW
+                dds.turns_to_ram(data, ram)
+                self.ftw = dds.frequency_to_ftw(base_frequency)
+                self.asf = dds.amplitude_to_asf(base_amplitude)
+            case RAMType.AMP:
+                self.dest = RAM_DEST_ASF
+                dds.amplitude_to_ram(data, ram)
+                self.ftw = dds.frequency_to_ftw(base_frequency)
+                self.pow = dds.turns_to_pow(base_phase)
+            case RAMType.POLAR:
+                self.dest = RAM_DEST_POWASF
+                phase, amp = data
+                dds.turns_amplitude_to_ram(phase, amp, ram)
+                self.ftw = dds.frequency_to_ftw(base_frequency)
+            case _:
+                raise NotImplementedError
+
+        self.start_addr = 0
+        self.end_addr = len(ram) - 1    # Inclusive
+
+        # Configure ther use of OSK. Note that using different OSK settings for
```

Comment:
Question: Is OSK used anywhere in the current ARTIQ/sinara code? Should we always disable OSK?

Reply:
OSK is exposed in ARTIQ but never used. We need to enable OSK if we ever use RAM for frequency/phase while having controllable amplitude.

If controlling amplitude on the fly is not needed, then I think we can just disable OSK. Though I need to check that ad9910 indeed does not get amplitude scaling from anywhere. (The second last entry on the data priority table of the amplitude column kind of worries me.)

Reply:
We definitely need to control the amplitude. I thought the amplitude is currently controlled by ASF, not OSK though. I probably did not understand it correctly.

Reply:
Single-tone profiles use the ASF field in the corresponding single-tone profile register (The `set()` function in `AD9910`). We don't need OSK for this.

But if we are using frequency/phase RAM, then we neither have amplitude data in RAM, nor ASF field in the profile registers. The simplest way to control amplitude is to use the ASF register, which needs OSK.

Reply:
I see. ASF register is different from the ASF in the single-tone profile register...